### PR TITLE
Refactored errors. Removed internal package.

### PR DESCRIPTION
**Pull request by yegorch**
* Created errors package and functions to inspect errors. 
* Removed internals. 
* Renamed endpoints to handlers

### CODE REVIEW

File: pkg/user/handler.go [hunk 1]
```diff
@@ -1,16 +1,15 @@
 package user
 
 import (
+	sgjwt "github.com/supergiant/supergiant/pkg/jwt"
+	"net/http"
```
Comment (saheienko) [inline pkg/user/handler.go]: Please, run 'make fmt'.

File: pkg/user/service.go [hunk 2]
```diff
@@ -1,61 +1,47 @@
...
 	"github.com/sirupsen/logrus"
-	"golang.org/x/crypto/bcrypt"
-
+	"github.com/supergiant/supergiant/pkg/sgerrors"
 	"github.com/supergiant/supergiant/pkg/storage"
+	"golang.org/x/crypto/bcrypt"
 )
 
-const prefix = "/user/"
+const prefix = "/supergiant/user/"
```
Comment (saheienko) [inline pkg/user/service.go]: Why do you hardcode the root prefix?

File: pkg/user/service.go [hunk 3]
```diff
@@ -1,61 +1,47 @@
...
-		}
-	}
-	logrus.Debug("user.Service.GetByToken end")
-	return nil, errors.Errorf("user with api token %s not found", apiToken)
-}
-
-func (s *Service) RegisterUser(ctx context.Context, user *User) error {
+func (s *Service) Create(ctx context.Context, user *User) error {
+	logrus.Debug("user.Service.Create start")
+	logrus.Debug("user.Service.Create end")
```
Comment (saheienko) [inline pkg/user/service.go]: ???

File: pkg/user/service.go [hunk 4]
```diff
@@ -1,61 +1,47 @@
...
 	return nil
 }
 
 func (s *Service) Authenticate(ctx context.Context, username, password string) error {
 	logrus.Debug("user.Service.Authenticate start")
 	rawJSON, err := s.Repository.Get(ctx, prefix, username)
 	if err != nil {
+		//If user doesn't exists we still want Forbidden instead of Not Found
+		if sgerrors.IsNotFound(err) {
+			return sgerrors.ErrInvalidCredentials
```
Comment (saheienko) [inline pkg/user/service.go]: Why do you return an ErrInvalidCredentials when user isn't found?

File: pkg/helm/repository/handler.go [hunk 5]
```diff
@@ -52,35 +53,40 @@ func (h *Handler) Get(w http.ResponseWriter, r *http.Request) {
...
 	repo, err := h.svc.Get(r.Context(), repoName)
 	if err != nil {
-		http.Error(w, "", http.StatusInternalServerError)
-		return
-	}
-	if repo == nil {
-		w.WriteHeader(http.StatusNotFound)
+		if sgerrors.IsNotFound(err) {
+			http.Error(w, err.Error(), http.StatusNotFound)
+		}
```
[the remark this answers is not in the document]
Comment (yegorch) [reply pkg/helm/repository/handler.go]: This is not bare error but a constant error message

[the remark this answers is not in the document]
Comment (saheienko) [reply pkg/helm/repository/handler.go]: err.Error() ?

File: pkg/helm/repository/handler.go [hunk 6]
```diff
@@ -52,35 +53,40 @@ func (h *Handler) Get(w http.ResponseWriter, r *http.Request) {
...
 
 // Delete removes a helm repository from the storage.
 func (h *Handler) Delete(w http.ResponseWriter, r *http.Request) {
 	repoName := mux.Vars(r)["repoName"]
 
-	err := h.svc.Delete(r.Context(), repoName)
-	if err != nil {
-		http.Error(w, "", http.StatusInternalServerError)
+	if err := h.svc.Delete(r.Context(), repoName); err != nil {
+		logrus.Error(err)
```
[the remark this answers is not in the document]
Comment (yegorch) [reply pkg/helm/repository/handler.go]: Fixed

File: pkg/kube/handler.go [hunk 7]
```diff
@@ -5,79 +5,74 @@ import (
...
 	vars := mux.Vars(r)
 	kubeId := vars["id"]
-
 	kube, err := e.service.Get(r.Context(), kubeId)
-
 	if err != nil {
-		http.Error(w, err.Error(), http.StatusNotFound)
+		if sgerrors.IsNotFound(err) {
+			http.Error(w, sgerrors.ErrNotFound.Error(), http.StatusNotFound)
+		}
```
[the remark this answers is not in the document]
Comment (yegorch) [reply pkg/kube/handler.go]: Fixed.

File: pkg/storage/storage.go [hunk 8]
```diff
@@ -20,51 +21,49 @@ type ETCDRepository struct {
 	cfg clientv3.Config
 }
 
-var ErrKeyNotFound = errors.New("key not found")
-
 func (e *ETCDRepository) Get(ctx context.Context, prefix string, key string) ([]byte, error) {
 	cl, err := e.GetClient()
 	if err != nil {
-		return nil, errors.WithStack(err)
+		return nil, errors.Wrap(err, "failed to connect to the etcd")
```
[the remark this answers is not in the document]
Comment (yegorch) [reply pkg/storage/storage.go]: Will fix later, as we were talking about refactoring this package anyway.

File: pkg/user/service.go (hunk 2 above)
Comment (yegorch) [reply pkg/user/service.go]: Will fix in later pull requests

File: pkg/kube/handler.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (saheienko) [reply pkg/kube/handler.go]: Still the same.

File: pkg/message/message.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	msg := New("Validation Failed", err.Error(), sgerrors.ValidationFailed, "")
+
+	data, err := json.Marshal(msg)
+	if err != nil {
+		logrus.Errorf("failed to marshall message: %v", err)
+		http.Error(w, "", http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
```
Comment (saheienko) [inline pkg/message/message.go]: A lot of the boilerplate code. Could you use a generic method for it?

File: pkg/message/message.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+func New(userMessage string, devMessage string, code sgerrors.ErrorCode, moreInfo string) Message {
+	return Message{
+		UserMessage: userMessage,
+		DevMessage:  devMessage,
+		ErrorCode:   code,
+		MoreInfo:    moreInfo,
+	}
+}
+func SendMessage(w http.ResponseWriter, msg Message, status int) {
+	data, err := json.Marshal(msg)
```
Comment (saheienko) [inline pkg/message/message.go]: Please, write tests for it.
You cant marshal a struct, it should be a pointer..

Comment (yegorch) [reply pkg/message/message.go]: @saheienko You can marshall a value, but it's have to be a pointer if you want to unmarshal it.

Comment (saheienko) [reply pkg/message/message.go]: Yeap, my fault.

**Pull request merged**